File: lib/JFFUtils/Dictionary/NSDictionary+BlocksAdditions.m
#import "NSDictionary+BlocksAdditions.h"

@implementation NSDictionary (BlocksAdditions)

- (instancetype)map:(JFFDictMappingBlock)block
{
    NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithCapacity:[self count]];
    [self enumerateKeysAndObjectsUsingBlock:^(id key, id object, BOOL *stop) {
        id newObject = block(key, object);
        NSParameterAssert(newObject);
        result[key] = newObject;
    }];
    return [result copy];
}

- (instancetype)forceMap:(JFFDictMappingBlock)block
{
    NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithCapacity:[self count]];
    [self enumerateKeysAndObjectsUsingBlock:^(id key, id object, BOOL *stop) {
        id newObject = block(key, object);
        if (newObject)
            result[key] = newObject;
    }];
    return [result copy];
}

- (instancetype)map:(JFFDictMappingWithErrorBlock)block outError:(NSError *__autoreleasing *)outError
{
    __block NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithCapacity:[self count]];
    
    __block NSError *error;
    
    [self enumerateKeysAndObjectsUsingBlock:^(id key, id object, BOOL *stop) {
        id newObject = block(key, object, &error);
        
        if (!newObject) {
            *stop = YES;
            result = nil;
            return;
        }
        
        result[key] = newObject;
    }];
    
    if (outError)
        *outError = error;
    
    return [result copy];
}

- (instancetype)mapKey:(JFFDictMappingBlock)block
{
    NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithCapacity:[self count]];
    [self enumerateKeysAndObjectsUsingBlock:^(id key, id object, BOOL *stop) {
        id newKey = block(key, object);
        NSParameterAssert(newKey);
        result[newKey] = object;
    }];
    return [result copy];
}

- (NSUInteger)count:(JFFDictPredicateBlock)predicate
{
    __block NSUInteger count = 0;
    [self enumerateKeysAndObjectsUsingBlock:^(id key, id object, BOOL *stop) {
        if (predicate(key, object))
            ++count;
    }];
    return count;
}

- (void)each:(JFFDictActionBlock)block
{
    [self enumerateKeysAndObjectsUsingBlock:^(id key, id object, BOOL *stop) {
        block(key, object);
    }];
}

- (instancetype)filter:(JFFDictPredicateBlock)predicate
{
    NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithCapacity:[self count]];
    
    [self enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
        
        if (predicate(key, obj)) {
            
            result[key] = obj;
        }
    }];
    
    return [result copy];
}

@end

File: iAsync.Swift/lib/JUtils/JUtils/Dictionary/NSDictionary+BlocksAdditions.swift
//
//  NSDictionary+BlocksAdditions.swift
//  JUtils
//
//  Created by Vladimir Gorbenko on 07.06.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

public extension NSDictionary {
    
    func map(block: JDictMappingBlock) -> NSDictionary {
        
        let result = NSMutableDictionary(capacity : count)
        
        enumerateKeysAndObjectsUsingBlock({(key : AnyObject!, object : AnyObject!, stop : UnsafeMutablePointer<ObjCBool>) -> Void in
            
            let newObject : AnyObject = block(key: key, object: object)
            result[key as! NSCopying] = newObject
        })
        
        return result.copy() as! NSDictionary
    }
    
    func forceMap(block: JDictOptionMappingBlock) -> NSDictionary {
        
        let result = NSMutableDictionary(capacity : count)
        
        enumerateKeysAndObjectsUsingBlock({(key : AnyObject!, object : AnyObject!, stop : UnsafeMutablePointer<ObjCBool>) -> Void in
            
            let newObject : AnyObject? = block(key : key, object : object)
            if let value : AnyObject = newObject {
                result[key as! NSCopying] = value
            }
        })
        
        return result.copy() as! NSDictionary
    }
    
    func map(block: JDictMappingWithErrorBlock, outError: NSErrorPointer) -> NSDictionary? {
        
        var result: NSMutableDictionary? = NSMutableDictionary(capacity: count)
        
        enumerateKeysAndObjectsUsingBlock({(key: AnyObject!, object: AnyObject!, stop: UnsafeMutablePointer<ObjCBool>) -> Void in
            
            let newObject : AnyObject? = block(key: key, object: object, outError: outError)
            
            if let value : AnyObject = newObject {
                
                result![key as! NSCopying] = value
            } else {
                
                if stop != nil {
                    stop.memory = true
                }
                result = nil
            }
        })
        
        return result?.copy() as? NSDictionary
    }
    
    func mapKey(block: JDictMappingBlock) -> NSDictionary {
        
        let result = NSMutableDictionary(capacity : count)
        
        enumerateKeysAndObjectsUsingBlock({(key : AnyObject!, object : AnyObject!, stop : UnsafeMutablePointer<ObjCBool>) -> Void in
            
            let newKey : AnyObject = block(key: key, object: object)
            //NSParameterAssert(newKey)
            result[newKey as! NSCopying] = object
        })
        
        return result.copy() as! NSDictionary
    }
    
    func count(predicate: JDictPredicateBlock) -> Int {
        
        var count = 0
        
        enumerateKeysAndObjectsUsingBlock({(key : AnyObject!, object : AnyObject!, stop : UnsafeMutablePointer<ObjCBool>) -> Void in
            if predicate(key: key, object: object) {
                ++count
            }
        })
        return count
    }
    
    func filter(predicate: JDictPredicateBlock) -> NSDictionary {
        
        let result = NSMutableDictionary(capacity: count)
        
        enumerateKeysAndObjectsUsingBlock({(key : AnyObject!, object : AnyObject!, stop : UnsafeMutablePointer<ObjCBool>) -> Void in
            
            if predicate(key: key, object: object) {
                result[key as! NSCopying] = object
            }
        })
        
        return result.copy() as! NSDictionary
    }
}
